File: 22/22.py
import re, sys, copy, random

# name, cost, damage, heal, manaCharge, armor, effectLength
spells = [
    ('MagicMissle',      53, 4, 0, 0, 0, 0),
    ('Drain',            73, 2, 2, 0, 0, 0),
    ('Shield',           113, 0, 0, 0, 7, 6),
    ('Poison',           173, 0, 0, 0, 0, 6),
    ('Recharge',         229, 0, 0, 101, 0, 5),
]
minspellcost = min([c[1] for c in spells])

class state(object):
    def __init__(self, player, boss, effects={}, manaspent=0, playerturn=True):
        self.player = copy.deepcopy(player)
        self.boss = copy.deepcopy(boss)
        self.effects = copy.deepcopy(effects)
        self.manaspent = manaspent
        self.playerturn = playerturn

    def canCast(self, spell):
        '''' Can the player cast the specified spell at this time?'''
        if spell[0] not in self.effects:
            if spell[1] <= self.player['mana']:
                return True
        return False

    def spellsAvailable(self):
        return [c for c in spells if self.canCast(c)]

    def cast(self, spell):
        # print spell
        self.player['mana'] -= spell[1]
        self.manaspent += spell[1]
        # activate effect
        if spell[6] > 0:
            self.effects[spell[0]] = spell[6]
        # apply damage to boss
        if spell[2] > 0:
            self.boss['hp'] -= spell[2]
        # apply healing to player
        if spell[3] > 0:
            self.player['hp'] += spell[3]

    def __str__(self):
        return ''.join(map(str, (self.player, self.boss, self.effects, self.playerturn)))

def runEffects(thisState):
    # apply active effects and decrement timers
    if 'Shield' in thisState.effects:
        thisState.player['armor'] = 7
        thisState.effects['Shield'] -=1
        if thisState.effects['Shield'] == 0:
            del thisState.effects['Shield']
    if 'Poison' in thisState.effects:
        thisState.boss['hp'] -= 3
        thisState.effects['Poison'] -=1
        if thisState.effects['Poison'] == 0:
            del thisState.effects['Poison']
    if 'Recharge' in thisState.effects:
        thisState.player['mana'] += 101
        thisState.effects['Recharge'] -=1
        if thisState.effects['Recharge'] == 0:
            del thisState.effects['Recharge']

    # check for end conditions
    if thisState.player['hp'] <= 0 or thisState.player['mana'] < minspellcost:
        # loss
        # print player,boss
        # print 'loss'
        return float('inf')
    if thisState.boss['hp'] <= 0:
        # win
        # print thisState.manaspent
        return thisState.manaspent

    return 0

def battle(thisState, winCost=float('inf')):
    ''' Fight until player dies, boss dies, or player mana is all spent.'''
    while True:
        # execute effects at start of turn
        result = runEffects(thisState)
        if result:
            # game over
            if result < winCost:
                # print winCost
                return result
            else:
                return winCost

        if thisState.playerturn:
            # execute player turn
            thisState.playerturn = False
            # cast a random spell
            thisState.cast(random.choice(thisState.spellsAvailable()))

        else:
            # boss turn
            thisState.playerturn = True
            thisState.player['hp'] -= max(1, thisState.boss['dmg'] - thisState.player['armor'])
            if thisState.player['hp'] <= 0:
                # game over
                break

    return winCost

if __name__ == '__main__':
    # 1. get puzzle input
    boss = {'hp':None, 'dmg':None, 'armor':0}
    with open(sys.argv[1], 'r') as f:
        for line in f.readlines():
            m = re.match(r'Hit Points: ([0-9]+)', line)
            if m:
                boss['hp'] = int(m.group(1))
            m = re.match(r'Damage: ([0-9]+)', line)
            if m:
                boss['dmg'] = int(m.group(1))
            m = re.match(r'Armor: ([0-9]+)', line)
            if m:
                boss['armor'] = int(m.group(1))

    player = {'hp':50, 'mana':500, 'armor':0}
    manaspent = float('inf')
    for i in xrange(1000000):
        s = state(player, boss)
        w = battle(s)
        if w < manaspent:
            print w
            manaspent = w

    print 'Answer to part 1: ' + str(manaspent)




    # print 'Answer to part 2: ' + ''

File: 22/22.rb
require 'active_support/all'

MINE_HP = 50
MINE_MANA = 500
BOSS_HP = 71
BOSS_DAMAGE = 10

# MINE_HP = 10
# MINE_MANA = 250
# BOSS_HP = 14
# BOSS_DAMAGE = 8

# battle class
class Battle
  attr_accessor :hp, :mana, :boss_hp, :effects, :mana_used, :result, :stack, :use

  def starting
    self.hp = MINE_HP
    self.mana = MINE_MANA
    self.boss_hp = BOSS_HP
    self.effects = {}
    self.mana_used = 0
    self.result = nil
    self.stack = []
    self.use = []
  end

  def boss_damage
    BOSS_DAMAGE
  end

  def armor
    result = 0
    effects.each do |_key, effect|
      result += effect[:armor] if effect[:armor].present?
    end
    result
  end

  def can_cast?(name, spell)
    return false if use.first.present? && use.first != name
    return false if mana < spell[:cost]
    return false if effects[name].present?
    true
  end

  def cast(name, spell)
    use.shift if use.present?
    stack << name
    self.mana -= spell[:cost]
    self.mana_used += spell[:cost]
    self.boss_hp -= spell[:damage] if spell.key?(:damage)
    update_result
    self.hp += spell[:heal] if spell.key?(:heal)
    effects[name] = spell[:effect].dup if spell[:effect].present?
  end

  def update_result
    self.result ||= :won if boss_hp <= 0
    self.result ||= :lost if hp <= 0
  end

  def turn
    effects.each do |_key, effect|
      self.boss_hp -= effect[:damage] if effect[:damage].present?
      update_result
      self.mana += effect[:mana] if effect[:mana].present?
      effect[:turns] -= 1
    end
    effects.delete_if { |_name, effect| effect[:turns] == 0 }
  end

  def boss_turn
    turn
    self.hp -= [1, boss_damage - armor].max
    update_result
  end

  def penalty
    self.hp -= 1
    update_result
  end

  def dup
    duplicate = self.class.new
    duplicate.hp = hp
    duplicate.mana = mana
    duplicate.boss_hp = boss_hp
    duplicate.effects = effects.deep_dup
    duplicate.mana_used = mana_used
    duplicate.result = result
    duplicate.stack = stack.dup
    duplicate.use = use.dup
    duplicate
  end
end

@spells = {}
@spells[:missile] = { cost: 53, damage: 4 }
@spells[:drain] = { cost: 73, damage: 2, heal: 2 }
@spells[:shield] = { cost: 113, effect: { turns: 6, armor: 7 } }
@spells[:poison] = { cost: 173, effect: { turns: 6, damage: 3 } }
@spells[:recharge] = { cost: 229, effect: { turns: 5, mana: 101 } }

@min_mana_kill = nil
@min_mana_stack = nil

def self.recursive(battle)
  battle.penalty
  battle.turn
  return if won_test(battle)
  return if battle.result == :lost
  return if @min_mana_kill.present? && @min_mana_kill < battle.mana_used

  @spells.each do |key, spell|
    next unless battle.can_cast?(key, spell)
    # print "#{key}\n"
    new_battle = battle.dup
    new_battle.cast(key, spell)
    new_battle.boss_turn
    recursive(new_battle)
  end
end

def self.won_test(battle)
  return false unless battle.result == :won
  if @min_mana_kill.nil? || @min_mana_kill > battle.mana_used
    @min_mana_kill = battle.mana_used
    @min_mana_stack = battle.stack
  end
  true
end

battle = Battle.new
battle.starting
# battle.use = [:recharge, :shield, :drain, :poison, :missile]

recursive(battle)
print "#{@min_mana_stack}\n"
print "#{@min_mana_kill}\n"
